Fix infinite call loop on theme change.

ajreckof opened the pull request:
Fixes #91553
Based on [this comment](https://github.com/godotengine/godot/pull/75482#discussion_r1582862095) I had removed `updating_theme` but it seems it is still necessary even with `bulk_theme_override`.

Chaosus commented:
Works fine but when switching theme not all elements switching correctly (I don't observe that on 4.3 dev 6):

![изображение](https://github.com/godotengine/godot/assets/3036176/80a7b4b0-67ce-4ce5-b45b-c8645532ab9f)

Chaosus reviewed (approved):
Perhaps it would be helpful as a quick fix.

ajreckof commented:
> Works fine but when switching theme not all elements switching correctly (I don't observe that on 4.3 dev 6):
> 
> ![изображение](https://private-user-images.githubusercontent.com/3036176/328037398-80a7b4b0-67ce-4ce5-b45b-c8645532ab9f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQ5MzQxNzYsIm5iZiI6MTcxNDkzMzg3NiwicGF0aCI6Ii8zMDM2MTc2LzMyODAzNzM5OC04MGE3YjRiMC02N2NlLTRjZTUtYjQ1Yi1jODY0NTUzMmFiOWYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDUwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA1MDVUMTgzMTE2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWRmZDFjN2M3NzA0MmZlYTNmZTJiOGQwNTU2MGI5NThkM2FlNzc4OTcwYjc5ZTk1YjExNGU3NmUxYTM0NGM2YiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.SqwKeKZp-QewlDlOpK1mhZiN5fZaUjWNb6y13kXSb_g)

Not related to the crash but fixed it as it was also introduced by my modification. While testing realized that `editorPropertyCategory` background is not updated on theme change, tested and it happened even before my modification so left it for later/ maybe there is already a issue / PR for it.

KoBeWi commented:
Yeah so the reason this is necessary is that EditorPropertyResource adds theme overrides inside its own `NOTIFICATION_THEME_CHANGED` and adding overrides will emit that notification again, resulting in infinite recursion.

I don't think it's valid to set theme overrides inside `NOTIFICATION_THEME_CHANGED` of the same object. The node could connect to `theme_changed` signal of parent inspector instead, to avoid calling theme changes on itself.

KoBeWi commented on editor/scene_tree_dock.cpp:
```diff
@@ -4354,7 +4354,6 @@ SceneTreeDock::SceneTreeDock(Node *p_scene_root, EditorSelection *p_editor_selec
 	scene_tree->set_v_size_flags(SIZE_EXPAND | SIZE_FILL);
 	scene_tree->connect("rmb_pressed", callable_mp(this, &SceneTreeDock::_tree_rmb));
 
-	scene_tree->connect("node_selected", callable_mp(this, &SceneTreeDock::_node_selected), CONNECT_DEFERRED);
```
What is this change for?

ajreckof replied:
I pushed a fix to the wrong branch sorry :(

ajreckof committed:
fix infinite call loop on theme change.

Apply suggestions from code review

ajreckof commented:
I updated it to use the parent `theme_changed` signal and since it was becoming too big to do separetly on each `EditorProperty` needing it I implemented it with a simple bool indicating wether it should have borders. it is only implemented to be set at init time. Should I comment on it or should I just make it work at any time with a dedicated setter?

KoBeWi commented on editor/editor_inspector.cpp:
```diff
@@ -405,6 +405,17 @@ void EditorProperty::_notification(int p_what) {
 		} break;
+		case NOTIFICATION_ENTER_TREE: {
+			if (has_borders) {
+				get_parent()->connect("theme_changed", callable_mp(this, &EditorProperty::_update_property_bg));
+				_update_property_bg();
+			}
+		} break;
+		case NOTIFICATION_EXIT_TREE: {
+			if (has_borders) {
+				get_parent()->disconnect("theme_changed", callable_mp(this, &EditorProperty::_update_property_bg));
```
```suggestion
				get_parent()->connect(SNAME("theme_changed"), callable_mp(this, &EditorProperty::_update_property_bg));
				_update_property_bg();
			}
		} break;
		case NOTIFICATION_EXIT_TREE: {
			if (has_borders) {
				get_parent()->disconnect(SNAME("theme_changed"), callable_mp(this, &EditorProperty::_update_property_bg));
```

ajreckof replied:
I'm assuming this is what you meant as the code you suggested didn't seem to work `SceneStringNames::get_singleton()->theme_changed`. Please correct me if this was not what you were suggesting.

AThousandShips replied:
No I suggested exactly what I suggested, it requires a rebase

ajreckof replied:
ok it was merged two hours ago that is why I didn't have it. Thanks for clarifying.

KoBeWi commented:
Property backgrounds don't refresh:
![image](https://github.com/godotengine/godot/assets/2223172/026d7e34-c886-4ab5-b5cc-9fe76213b81a)

ajreckof commented:
Fixed thanks

akien-mga commented:
Thanks!
